library/alloc/src/sync.rs

```diff
@@ -943,11 +949,16 @@ impl<T: ?Sized> Arc<T> {
         let mut cur = this.inner().weak.load(Relaxed);
 
         loop {
-            // check if the weak counter is currently "locked"; if so, spin.
-            if cur == usize::MAX {
-                hint::spin_loop();
-                cur = this.inner().weak.load(Relaxed);
-                continue;
+            if cur > MAX_REFCOUNT {
+                // check if the weak counter is currently "locked"; if so, spin.
+                if cur == usize::MAX {
+                    hint::spin_loop();
+                    cur = this.inner().weak.load(Relaxed);
+                    continue;
+                }
+                // We shouldn't allow the refcount to increase much past `MAX_REFCOUNT`.
+                // See https://github.com/rust-lang/rust/issues/108706
+                panic!("{}", INTERNAL_OVERFLOW_ERROR);
             }
```

Comment:
Instead of a nested `if`, maybe it's a bit easier to read to leave the `if cur == MAX` block untouched, and just add a single `assert!(cur <= MAX_REFCOUNT, "...")` afterwards.

Reply:
I went ahead and changed this in `upgrade` as well.